### PR TITLE
enhancement(kafka source): Add support for acknowledgements

### LINKED ISSUE #7460: Add support for end-to-end acknowledgements to `kafka` source
The `kafka` source needs to wait for a delivery acknowledgement before marking a batch of events as committed. This is complicated by the possibility of the events being acknowledged out-of-order but commits have to happen in order as an "offset" value. Some work will need to go into tracking this high water mark internally.

Ref #7336

### PR DESCRIPTION
**Pull request by bruceg**
There are a couple of warts left with this, but it is feature complete.

Closes #7460

### CODE REVIEW

File: src/sources/kafka.rs [hunk 1]
```diff
@@ -218,6 +233,99 @@ async fn kafka_source(
...
+        let mut result = None;
+        while let Entry::Occupied(entry) = self.map.entry(next) {
+            if !entry.get().done {
+                break;
+            }
+            let entry = entry.remove();
+            result
+                .get_or_insert_with(TopicPartitionList::new)
+                .add_partition(&entry.topic, entry.partition)
+                .set_offset(Offset::from_raw(entry.offset + 1)) // FIXME why does the +1 make it work?
```
Comment (bruceg) [inline src/sources/kafka.rs:274]: When I just use `Offset::from_raw(entry.offset)`, this fails with an "invalid parameter" error. Adding the `+ 1` makes it work, and produces the same high water mark as the non-acknowledgement code path. I have no idea why.

File: src/sources/kafka.rs [hunk 2]
```diff
@@ -392,28 +518,36 @@ mod integration_test {
...
-        assert!(events[0].as_log().contains("partition"));
-        assert!(events[0].as_log().contains("offset"));
-        let mut expected_headers = BTreeMap::new();
-        expected_headers.insert("my header".to_string(), Value::from("my header value"));
-        assert_eq!(events[0].as_log()["headers"], Value::from(expected_headers));
+        let events = collect_n(rx, 10).await;
+
+        let client: BaseConsumer = client_config();
+        let watermarks = client
+            .fetch_watermarks(&topic, 0, Duration::from_secs(1))
```
Comment (bruceg) [inline src/sources/kafka.rs:527]: This sometimes dies with an "invalid partition" error. However, shouldn't there only be one partition (numbered `0`)? Any ideas what's going on there?

**Comment by bruceg** [comment]
While the tests pass, I have confirmed with debug print statements that acknowledgements are not getting to kafka when they are enabled. Also, the watermarks test is not verifying what I thought it should, so this still needs some work.

**Comment by bruceg** [comment]
This will be reworked once the file source acks have landed (#7816) to reuse the `OrderedFinalizer` introduced there.

**Comment by bruceg** [comment]
Ok, I have resolved the behavior issues with this and rebased it on top of the file source work, so this is good for a review again.

**Comment by bruceg** [comment]
Started integration test run: https://github.com/timberio/vector/actions/runs/948354883

**Review by lukesteensen — APPROVED** [review]

**Pull request merged**